### PR TITLE
Fixed issue #2024, can now remove line and block comments with LexicalPreservingPrinter

### LINKED ISSUE #2024: LexicalPreservingPrinter throws UnsupportedOperationException when removing comments
Hi,
Whenever I try to delete a line or block comment while using the LexicalPreservingPrinter, I get an UnsupportedOperationException. Javadoc comments seem to work fine.

Example:
```
        FileInputStream in = new FileInputStream(path);
        CompilationUnit compilationUnit = LexicalPreservingPrinter.setup(JavaParser.parse(in));
		
		List<Comment> comments = compilationUnit.getAllContainedComments();
		
		for (Comment comment : comments) {
			// throws UnsupportedOperationException unless it's a Javadoc comment
			comment.remove();
		}
```

### PR DESCRIPTION
**Pull request by JustinKissling**
Fixes #2024 by modifying LexicalPreservingPrinter to handle removed block and line comments.
Also includes test cases.

### CODE REVIEW

File: javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java [hunk 1]
```diff
@@ -1105,6 +1105,46 @@ void addedLineCommentsPrinted() {
                           "}" + EOL +
                           "}", LexicalPreservingPrinter.print(cu));
     }
+
```
Comment (ftomassetti) [inline javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java:1104]: +1 for the tests

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java [hunk 2]
```diff
@@ -164,27 +170,43 @@ public void concretePropertyChange(Node observedNode, ObservableProperty propert
             LEXICAL_DIFFERENCE_CALCULATOR.calculatePropertyChange(nodeText, observedNode, property, oldValue, newValue);
         }
 
-        private List<TokenTextElement> getMatchingTokenTextElements(JavadocComment oldValue, NodeText nodeText) {
-            JavadocComment javadocComment = oldValue;
-            List<TokenTextElement> matchingTokens = nodeText.getElements().stream()
+        private List<TokenTextElement> getMatchingTokenTextElements(Comment oldValue, NodeText nodeText) {
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java]: should this method name contains the word comment?

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java [hunk 3]
```diff
@@ -164,27 +170,43 @@ public void concretePropertyChange(Node observedNode, ObservableProperty propert
...
+            List<TokenTextElement> matchingTokens;
+            
+            if (oldValue instanceof JavadocComment) {
+                matchingTokens = nodeText.getElements().stream()
                     .filter(e -> e.isToken(JAVADOC_COMMENT))
                     .map(e -> (TokenTextElement) e)
-                    .filter(t -> t.getText().equals("/**" + javadocComment.getContent() + "*/"))
+                    .filter(t -> t.getText().equals("/**" + oldValue.getContent() + "*/"))
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:180]: Do we have a utility method to produce this string?

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java [hunk 4]
```diff
@@ -164,27 +170,43 @@ public void concretePropertyChange(Node observedNode, ObservableProperty propert
...
+                    .collect(Collectors.toList());
+            } else if (oldValue instanceof BlockComment) {
+                matchingTokens = nodeText.getElements().stream()
+                    .filter(e -> e.isToken(MULTI_LINE_COMMENT))
+                    .map(e -> (TokenTextElement) e)
+                    .filter(t -> t.getText().equals("/*" + oldValue.getContent() + "*/"))
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:186]: And this

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java [hunk 5]
```diff
@@ -164,27 +170,43 @@ public void concretePropertyChange(Node observedNode, ObservableProperty propert
...
                     .collect(Collectors.toList());
+            } else {
+                matchingTokens = nodeText.getElements().stream()
+                    .filter(e -> e.isToken(SINGLE_LINE_COMMENT))
+                    .map(e -> (TokenTextElement) e)
+                    .filter(t -> t.getText().equals("//" + oldValue.getContent() + System.lineSeparator()))
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:192]: And this

File: javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java [hunk 6]
```diff
@@ -1105,6 +1105,46 @@ void addedLineCommentsPrinted() {
...
+        LexicalPreservingPrinter.setup(cu);
+        cu.getAllContainedComments().get(0).remove();
+
+        assertEqualsNoEol("public class Foo {" + EOL +
+                          "void mymethod() {" + EOL +
+                          "}" + EOL +
+                          "}", LexicalPreservingPrinter.print(cu));        
+        
+        /*
+
```
Comment (matozoid) [inline javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java]: Huh? :-)

**Review by ftomassetti — APPROVED** [review]
+1
Thank you for the great contribution!

**Comment by matozoid** [comment]
Thanks! This is good stuff :-)

File: javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java (hunk 6 above)
Comment (JustinKissling) [reply javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinterTest.java]: Ah, my bad. Removed now.

**Comment by matozoid** [comment]
@JustinKissling can you take a look at @ftomassetti's remarks?

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java (hunk 3 above)
Comment (JustinKissling) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:180]: I don't think there is a utility method for this. I saw this way of assembling the string used in other parts of the code (like the prettyPrintingTextNode method) so I simply did the same thing here.

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java (hunk 2 above)
Comment (JustinKissling) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java]: I renamed the method to "findTokenTextElementForComment" now, that should describe its purpose more accurately

**Comment by JustinKissling** [comment]
@matozoid Done. Let me know if I should do anything about the last three remarks. The way I did it is the same as in other parts of the code, but I can also add an extra method for getting this string from comments if needed

**Comment by matozoid** [comment]
@JustinKissling let's leave that for another time. Time to merge!

**Pull request merged**